setup.py

```diff
@@ -1,127 +1,214 @@
+
+
+def write_version_file(version, git_revision):
+    """
+    Write version information to the version file.
+
+    Overwrites any existing version file.
 
-    if not IS_RELEASED:
-        fullversion += '.dev{0}'.format(dev_num)
+    Parameters
+    ----------
+    version : unicode
+        Package version.
+    git_revision : unicode
+        The full commit hash for the current Git revision.
+    """
+    with io.open(VERSION_FILE, "w", encoding="ascii") as version_file:
+        version_file.write(
+            VERSION_FILE_TEMPLATE.format(
+                version=version, git_revision=git_revision
+            )
+        )
+
+
+def read_version_file():
+    """
+    Read version information from the version file, if it exists.
+
+    Returns
+    -------
+    version : unicode
+        The full version, including any development suffix.
+    git_revision : unicode
+        The full commit hash for the current Git revision.
+
+    Raises
+    ------
+    EnvironmentError
+        If the version file does not exist.
+    """
+    version_info = runpy.run_path(VERSION_FILE)
+    return (version_info["version"], version_info["git_revision"])
+
+
+def git_version():
+    """
+    Construct version information from local variables and Git.
+
+    Returns
+    -------
+    version : unicode
+        Package version.
+    git_revision : unicode
+        The full commit hash for the current Git revision.
+
+    Raises
+    ------
+    EnvironmentError
+        If Git is not available.
+    subprocess.CalledProcessError
+        If Git is available, but the version command fails (most likely
+        because there's no Git repository here).
+    """
+    git_count, git_revision = _git_info()
+    version_template = RELEASED_VERSION if IS_RELEASED else UNRELEASED_VERSION
+    version = version_template.format(
+        major=MAJOR, minor=MINOR, micro=MICRO, dev=git_count
+    )
+    return version, git_revision
+
+
+def resolve_version():
+    """
+    Process version information and write a version file if necessary.
+
+    Returns the current version information.
+
+    Returns
+    -------
+    version : unicode
+        Package version.
+    git_revision : unicode
+        The full commit hash for the current Git revision.
+    """
+    if os.path.isdir(GIT_DIRECTORY):
+        # This is a local clone; compute version information and write
+        # it to the version file, overwriting any existing information.
+        version = git_version()
+        print(u"Computed package version: {}".format(version))
+        print(u"Writing version to version file {}.".format(VERSION_FILE))
+        write_version_file(*version)
+    elif os.path.isfile(VERSION_FILE):
+        # This is a source distribution. Read the version information.
+        print(u"Reading version file {}".format(VERSION_FILE))
+        version = read_version_file()
+        print(u"Package version from version file: {}".format(version))
+    elif IS_RELEASED:
```

Comment:
Can we swap these two `elif` clauses around? It probably won't ever matter, but if this file has been updated to make `IS_RELEASED=True`, then the information in this file should take precedence over anything that happens to be in the `version.py` file.